Tools/MonoGame.Effect.Compiler/Effect/ShaderProfile.Vulkan.cs

```diff
@@ -262,148 +232,106 @@ internal override ShaderData CreateShader(ShaderResult shaderResult, string shad
+                    else if (variable.Pointer.PointerType.Type == SpirvType.Image)
                     {
-                        if (data[2] == "OpTypeStruct")
-                        {
-                            var id = data[0];
-                            var s = structs[id];
+                        // find all the times this image was sampled by distinct samplers.
+                        var sampledImages = reflectionInfo.SampledImages.Where(si => si.LoadedImage.Variable == variable)
+                            .DistinctBy(si => si.LoadedSampler.Variable);
 
-                            for (int i=3; i < data.Length; i++)
-                                s.members[i - 3].type = data[i];
-                        }
-                        else if (data[2] == "OpVariable")
+                        // and create a sampler parameter for each.
+                        foreach (var sampledImage in sampledImages)
                         {
-                            VkDescriptor d;
-                            if (descriptors.TryGetValue(data[0], out d))
+                            // pull out what we need from the sampled image object
+                            var samplerVariable = sampledImage.LoadedSampler.Variable;
+                            var imageVariable = sampledImage.LoadedImage.Variable;
+
+                            var samplerType = samplerVariable.Pointer.PointerType as SpirvTypeSampler;
+                            var imageType = imageVariable.Pointer.PointerType as SpirvTypeImage;
+
+                            var sampler = new ShaderData.Sampler
                             {
-                                if (data[3].EndsWith("_type_2d_image"))
-                                    d.type = VkDescriptorType.SAMPLED_IMAGE;
-                                else if (data[3].EndsWith("_type_sampler"))
-                                    d.type = VkDescriptorType.SAMPLER;
-                            }
-                        }
-                    }
-                }
+                                samplerSlot = samplerVariable.BindingSlot.Value - SlotOffset,
+                                samplerName = samplerVariable.Name,
+                                textureSlot = imageVariable.BindingSlot.Value - SlotOffset,
+                            };
 
-                var cbuffer = new ConstantBufferData("global");
+                            // This image is only sampled by one sampler, we can safely use the texture name for the parameter.
+                            if (sampledImages.Count() == 1)
+                            {
+                                sampler.parameterName = imageVariable.Name;
+                            }
+                            // otherwise make a composite name for this image/sampler combo.
+                            else
+                            {
+                                sampler.parameterName = $"{samplerVariable.Name}+{imageVariable.Name}";
```

Comment:
I do want to cleanup all the mess with sampler/texture issues and make it consistent across all target shader profiles.  This seems like a good way forward.